dexcom/data_range.go

```diff
@@ -0,0 +1,149 @@
+package dexcom
+
+import (
+	"time"
+
+	"github.com/tidepool-org/platform/structure"
+	structureValidator "github.com/tidepool-org/platform/structure/validator"
+)
+
+type DataRange struct {
+	Start *Times `json:"start,omitempty"`
+	End   *Times `json:"end,omitempty"`
+}
+
+type Times struct {
+	SystemTime  *Time `json:"systemTime,omitempty"`
+	DisplayTime *Time `json:"displayTime,omitempty"`
+}
+
+type DataRangeResponse struct {
+	RecordType    *string    `json:"recordType,omitempty"`
+	RecordVersion *string    `json:"recordVersion,omitempty"`
+	UserID        *string    `json:"userId,omitempty"`
+	Calibrations  *DataRange `json:"calibrations,omitempty"`
+	EGVs          *DataRange `json:"egvs,omitempty"`
+	Events        *DataRange `json:"events,omitempty"`
+}
+
+func ParseDataRangeResponse(parser structure.ObjectParser) *DataRangeResponse {
+	if !parser.Exists() {
+		return nil
+	}
+	datum := NewDataRangeResponse()
+	parser.Parse(datum)
+	return datum
+}
+
+func NewDataRangeResponse() *DataRangeResponse {
+	return &DataRangeResponse{}
+}
+
+func (d *DataRangeResponse) Parse(parser structure.ObjectParser) {
+	d.UserID = parser.String("userId")
+	d.RecordType = parser.String("recordType")
+	d.RecordVersion = parser.String("recordVersion")
+	d.Calibrations = ParseDataRange(parser.WithReferenceObjectParser("calibrations"))
+	d.EGVs = ParseDataRange(parser.WithReferenceObjectParser("egvs"))
+	d.Events = ParseDataRange(parser.WithReferenceObjectParser("events"))
+}
+
+func (d *DataRangeResponse) GetOldestStartDate() time.Time {
```

Comment:
I think GetOldestStartDate should return an error if none of the three time values are found.

```suggestion
func (d *DataRangeResponse) GetOldestStartDate() (time.Time, error) {
```

Right now, time.Now() is returned, then [it's paired with a minute ago](https://github.com/tidepool-org/platform/blob/back-325-data-range/dexcom/fetch/runner.go#L400), resulting in a negative range request to Dexcom, that even if they accept it and run it, is still requesting data for a span of time that's at most 1 minute. It's a waste of resources.

Instead, if none of these three values is found, we should return an error, and not waste more of our time, or Dexcom's.

The above should never happen anyway, as that would be an invalid DataRangeResponse, and should be caught before getting here, but why not just be safe and return an error?